apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/ValueConversion.cs

```diff
@@ -355,18 +343,72 @@ private static object ConvertToClrTypeImpl(Value wireValue, V1.Type spannerType,
                         return dictionary;
                     case Value.KindOneofCase.ListValue:
                         dictionary = new Hashtable();
-                        var i = 0;
-                        foreach (var listItemValue in wireValue.ListValue.Values)
+                        if (spannerType.Code == TypeCode.Struct)
                         {
-                            dictionary[i] = listItemValue.ConvertToClrType(spannerType.ArrayElementType);
-                            i++;
+                            for (int i = 0; i < spannerType.StructType.Fields.Count; i++)
+                            {
+                                dictionary[spannerType.StructType.Fields[i].Name] = wireValue
+                                    .ListValue.Values[i]
+                                    .ConvertToClrType(spannerType.StructType.Fields[i].Type);
+                            }
+                        }
+                        else
+                        {
+                            var i = 0;
+                            foreach (var listItemValue in wireValue.ListValue.Values)
+                            {
+                                dictionary[i] =
+                                    listItemValue.ConvertToClrType(spannerType.ArrayElementType);
+                                i++;
+                            }
                         }
                         return dictionary;
                     default:
                         throw new ArgumentException(
                             $"Invalid Type conversion from {wireValue.KindCase} to {targetClrType.FullName}");
                 }
             }
+            if (targetClrType.IsArray)
+            {
+                switch (wireValue.KindCase)
+                {
+                    case Value.KindOneofCase.NullValue: return null;
+                    case Value.KindOneofCase.ListValue:
+                        Array newArray = Array.CreateInstance(targetClrType.GetElementType(),
+                            wireValue.ListValue.Values.Count);
+
+                        int i = 0;
+                        foreach (var obj in wireValue.ListValue.Values.Select(
+                            x => x.ConvertToClrType(spannerType.ArrayElementType, targetClrType.GetElementType())))
+                        {
+                            newArray.SetValue(obj, i);
+                            i++;
+                        }
+                        return newArray;
+                    default:
+                        throw new InvalidOperationException(
+                            $"Invalid Type conversion from {wireValue.KindCase} to {targetClrType.FullName}");
+                }
+
+            }
+            if (typeof(IList).IsAssignableFrom(targetClrType))
```

Comment:
they might be able to be combined, this is for when the user passes in a concrete (possibly generic) type.  I think they could probably be merged together.